Implementation of moisture sensor

LegyPlegy opened the pull request:
I have updated the readme with information on `therm` and `moist`. `moist` code has full functionality, meaning that it correctly calls the method `sound_the_alarm` when submerged in water and does nothing when dry. I am not sure why the method seems to lag, though. It takes about 3 seconds to sound an alarm when it is submerged in water. Also, rapid testing where you repeatedly dip the sensor in and out of the water doesn't trigger the alarm, which may seem to be a problem.

I should write a note somewhere that both contacts on the leak sensor need to share the same puddle of water to work properly. The sensor operates by providing a trickle current between its `ground (white)` and `data (red)` wires when submerged in water, and is an open circuit otherwise. 

The code is configured for pin 11, using `BCM 17`, and it is opened as an `Input` with an internal PullUp resistor. I would test tomorrow to see if wiring the pullup resistor on the breadboard will help with faster responses, however the manual recommends a 5MOhm resistor and the highest one in the lab is 3MOhm, meaning that this is currently not possible.

ypsun-umd commented:
Ah ha. I like the idea of event detect. which avoids the dreaded `while True` loop. Still, you should try the (most) stupid `while True; sleep..` snippet in my `firealarm` folder to see if it still lags.

LegyPlegy commented:
Summary:

Finalized leaksensor.py, it is fully functional and works exceptionally. After revising the circuit I realized that for our purposes it makes much more sense to wire the leaksensor as an analog input into the GPIO. With the internal pulldown resistor in the RPI setting pin 11 to LOW, the script will monitor pin 11 for any voltage introduced by the sensor closing the circuit as a result of contact with water. Works as fast as the Pi can process, although I set a time.sleep(0.01) so it does not hog the processor. **I also modified the alarm.cleanup() method so that it only cleans the channels that it has opened, so that no other channels are closed accidentally.**

I began working with pathlib to automate the script detecting and saving the sensors that it detects. In doing so, I realized that the temperature sensor script is getting worse and worse to read, so later this week I will try and rewrite it in a much more readable format.


Also updated readme.md.

ypsun-umd commented:
Also for the temperature sensor, you used a `with` block. Note that there is a `while` loop within that block, which means the file is **never** closed before the program exits. I would suggest the following:
``` 
while True:
    with file as f:
        ...do stuff...
```

LegyPlegy commented:
Rewrote thermal sensor code, deleted auxiliary files. Updated readme to separate acknowledgments and documentation for each piece of code. Very brief, not sure if it would be desired to write in-depth for the documentation. 

Note that thermsensor.py has built-in functionality to allow the user to set a custom delay via arguments in the terminal. "python3 thermsensor.py 3" will run the script with a 3-second delay between measurements.

Also, @ypsun-umd I do not understand what you mean by the temperature sensor having an indefinitely opened `with` block. The `with` block is contained within the method `extract_therm`, and even though `extract_therm` is inside a `while True` loop, the `with` statement should close right before the method returns if I understand correctly. I would appreciate it if you could double check that for me, I'm not well-versed in the logic behind the `with` statement.

ypsun-umd commented:
I skimmed through, and now there's no problem with `with`. Because now `with` and `while` are **not** in the same block. 

That explanation is probably confusing. Feel free to ask next time you come around.

ypsun-umd commented:
@LegyPlegy will you come today (Aug 15)? If so, we can sit down together and I'll explain why I would edit your code to a certain way.
